x/zkauth/types/zkauth_inputs.go

```diff
@@ -0,0 +1,199 @@
+package types
+
+import (
+	"encoding/base64"
+	"math"
+	"math/big"
+	"unicode/utf8"
+
+	"github.com/iden3/go-iden3-crypto/poseidon"
+	"github.com/pkg/errors"
+
+	sdkerrors "github.com/Finschia/finschia-sdk/types/errors"
+)
+
+const MaxHeaderLen = 992
+const MaxExtIssLen = 32
+const PackWidth = 248
+
+// circom constants from main.circom / https://zkrepl.dev/?gist=30d21c7a7285b1b14f608325f172417b
+// template RSAGroupSigVerify(n, k, levels) {
+// component main { public [ modulus ] } = RSAVerify(121, 17)
+// component main { public [ root, payload1 ] } = RSAGroupSigVerify(121, 17, 30)
+
+const CircomBigintN = 121
+const CircomBigintK = 17
+
+func SplitToTwoFrs(ephPkBytes []byte) []*big.Int {
+	mid := len(ephPkBytes) - 16
+	first := new(big.Int).SetBytes(ephPkBytes[0:mid])
+	second := new(big.Int).SetBytes(ephPkBytes[mid:])
+
+	return []*big.Int{first, second}
+}
+
+func ToField(val string) (*big.Int, bool) {
+	return new(big.Int).SetString(val, 10)
+}
+
+func reverse[T any](s []T) {
+	for i, j := 0, len(s)-1; i < j; i, j = i+1, j-1 {
+		s[i], s[j] = s[j], s[i]
+	}
+}
+
+// ChunkArray splits an array into chunks of a specified size.
+func ChunkArray(array []byte, chunkSize int) [][]byte {
+	// The length of the array divided by the chunk size is rounded up and used as the number of chunks.
+	chunkCount := int(math.Ceil(float64(len(array)) / float64(chunkSize)))
+	chunks := make([][]byte, chunkCount)
+
+	revArray := make([]byte, len(array))
+	copy(revArray, array)
+	reverse(revArray)
+
+	// split the revert array into chunks, and revert each chunk again.
+	for i := 0; i < chunkCount; i++ {
+		start := i * chunkSize
+		end := start + chunkSize
+		if end > len(revArray) {
+			end = len(revArray)
+		}
+		chunk := make([]byte, end-start)
+		copy(chunk, revArray[start:end])
+		reverse(chunk)
+		chunks[i] = chunk
+	}
+
+	// return the final chunk array reverted.
+	reverse(chunks)
```

Comment:
By specifying the array range, it can be written more simply than in Javascript. The amount of calculation will also be reduced from O(n^2) to O(n) since reverse is eliminated.
```suggestion
	shouldFirstPack := chunkSize
	if len(array)%chunkSize != 0 {
		shouldFirstPack = len(array) % chunkSize
	}
	for i := 0; i < chunkCount; i++ {
		if i == 0 {
			chunks[i] = array[0:shouldFirstPack]
		} else {
			chunks[i] = array[shouldFirstPack+((i-1)*chunkSize) : shouldFirstPack+(i*chunkSize)]
		}
	}
```